text/0129-video-replay-envelope.md

```diff
@@ -0,0 +1,78 @@
+- Start Date: 2024-02-06
+- RFC Type: feature
+- RFC PR: [#129](https://github.com/getsentry/rfcs/pull/129)
+- RFC Status: draft
+
+# Summary
+
+In order to capture video replays and present them to the user, we need to define/decide:
+
+- how to transport the video data to the server
+- how to integrate the video data into the RRWeb JSON format
+- how to combine multiple replay chunks to a single replay session
+
+All of these influence one another and need to be considered together in a single RFC.
+
+Note: SDK-side implementation that is currently being worked on relies on taking screenshots and encoding them to a video.
+This is based on an evaluation where a video has much smaller size than a sequence of images (**TODO fix these: factor of X for 720p video**).
+
+# Motivation
+
+We need this to to capture replays on platforms where it's not possible/feasible to produce an HTML DOM (i.e. the native format supported by RRWeb). For example: mobile apps.
+
+<!-- # Supporting Data -->
+<!-- Metrics to help support your decision (if applicable). -->
+
+# Options Considered
+
+## Using a video, with EnvelopeItem:ReplayVideo
+
+- From the SDK, we would send a new envelope item type, [`ReplayVideo`](https://github.com/getsentry/relay/blob/5fd3969e88d3eea1f2849e55b61678cac6b14e44/relay-server/src/envelope.rs#L115C5-L115C20) to transport the video data.
+  The envelope item would consist of a single header line (JSON), followed by a new line and the raw video data.
```

Comment:
Are these values gonna be used, or they are here "just in case"? I think we should spec them out as well, if they are gonna be needed on backend/frontend

Reply:
Segment-id is needed in the header so I know what to name the file.  All the other head examples should be moved to the RRWeb JSON and can be kept just-in-case.

Reply:
i've updated the markdown in the meantime

Reply:
It's currently there in the updated json header example

Reply:
thanks! @cmanallen I've checked the item headers once again, e.g. for an attachment they look like that:
```json
{"type":"attachment","length":10,"content_type":"text/plain","filename":"hello.txt"}\n
```
does it make sense to adhere to the common pattern here and do something like this (`type` is actually required afaik):
```json
{"type":"replay_video","length":1000,"content_type":"video/mp4","segment_id":4}\n
```
`filename` is probably not needed if we name it by the segment_id, right? (4.mp4)

Reply:
> filename is probably not needed if we name it by the segment_id, right?

Correct.

> does it make sense to adhere to the common pattern here and do something like this (type is actually required afaik):

No I don't think its necessary to follow that pattern.  They accept those headers because they want to store them.  I have no intention of storing that metadata anywhere except in the RRWeb JSON.  If there's something about these headers you'd like to preserve, you should put it in the RRWeb JSON.  You can put anything you want there; it doesn't concern me at all.  I would recommend putting more than you need in there.  It can always be removed later.

Reply:
these are not part of the envelope item content (what is described here), but part of the envelope item specification.

Reply:
> The envelope item would consist of a single header line (JSON)

but this is the item header or am I wrong?

Reply:
I don't think that's what the ReplayRecording header does. AFAIU It is part of the envelope item payload. I'm not saying that's the right way to do it or not, just how I understood it is done for that envelope item and proposed to do the same for the new one.

Reply:
But I may be wrong, of course :)

Reply:
ah, got it now, thanks. So we just follow the usual pattern for the item header then (I guess we don't need to document it here, or do we?).

Reply:
> I guess we don't need to document it here, or do we?

It evidently is confusing so I'll clarify